crossplane/function/resource.py

```diff
@@ -17,12 +17,41 @@
 import dataclasses
 import datetime
 
+import pydantic
 from google.protobuf import struct_pb2 as structpb
 
+import crossplane.function.proto.v1.run_function_pb2 as fnv1
+
 # TODO(negz): Do we really need dict_to_struct and struct_to_dict? They don't do
 # much, but are perhaps useful for discoverability/"documentation" purposes.
 
 
+def update(r: fnv1.Resource, source: dict | structpb.Struct | pydantic.BaseModel):
+    """Update a composite or composed resource.
+
+    Use update to add or update the supplied resource. If the resource doesn't
+    exist, it'll be added. If the resource does exist, it'll be updated. The
+    update method semantics are the same as a dictionary's update method. Fields
+    that don't exist will be added. Fields that exist will be overwritten.
+
+    The source can be a dictionary, a protobuf Struct, or a Pydantic model.
+    """
+    # If the resource has a model_dump attribute it's a Pydantic BaseModel.
+    if hasattr(source, "model_dump"):
```

Comment:
@bobh66 Let me know if there's a more Pythonic way to determine the type of the `source` argument. I'm still pretty rusty. 😄

Reply:
I believe `isinstance` would be how to determine if it's a subclass

Reply:
@RedbackThomson Oh duh, good point.

Reply:
I knew Python recently got a `match` statement which is kinda like a Go switch. A bit of Googling led me to find that they now have a 'class match', which is similar to a Go type switch. I've updated the code to use that.